docs/secure-coding-guidelines.md

```diff
@@ -0,0 +1,172 @@
+# Secure Coding Guidelines
+
+## Purpose
+
+The secure coding guidelines are a reference to develop, deliver and maintain software in a secure manner.
+
+## Scope
+
+These guidelines apply to developers building client based applications at MetaMask.
+
+## References
+
+The guidelines in this policy were gathered primarily from the [OWASP Top 10](https://owasp.org/www-project-top-ten/) and the [OWASP Application Security Verification Standard](https://owasp.org/www-project-application-security-verification-standard/). We have included just the guidelines most relevant to MetaMask client applications.
+
+## Guidelines
+
+### Input from Users and Third Party Services
+
+#### Data Validation and Sanitization
+
+- Validate and sanitize all user input, and all data from external sources
+
+  For example:
+
+  - Check that the type matches expectations
+    - If a value should be a string, do not allow it to be an object
+    - Validation libraries can be used for more complex types of data
+  - Check that the value is is an allowed value or within the expected range
+    - If only a fixed set of values are expected, ensure the value matches one of them
+    - If a non-negative number is expected, do not allow the value to be negative
+  - Check that the format matches expectations
+    - If we expect a 0x-prefixed hexadecimal string, ensure that the 0x is present
+
+- Encode data before output
+
+  For example:
+
+  - HTML-encode strings before they are included in the DOM
+    - Some libraries do this automatically (e.g. React)
+  - URL-encode data before including it in a URL
+
+- When accepting a URI as input, ensure the scheme matches expectations
+
+  For example, a URL for a website or API would typically have a scheme of `https`
+
+- Avoid dynamic code execution with untrusted data
```

Comment:
Do you think we should add a statement? 

If dynamic code execution is a required part of a feature, a threat model and Security Team review should occur.

Reply:
Makes sense, I'll try to work this in

Reply:
Done here: 3f7883e584889119ca21c85d7afbc071b7015d7d